src/include/function/gds/output_writer.h

```diff
@@ -0,0 +1,136 @@
+
+    void beginFrontierComputeBetweenTables(common::table_id_t,
+        common::table_id_t nextFrontierTableID) override {
+        // Note: We do not fix the node table for pathLengths, which is inherited from AllSPOutputs.
+        // See the comment in SingleSPOutputs::beginFrontierComputeBetweenTables() for details.
+        bfsGraph.pinNodeTable(nextFrontierTableID);
+    };
+
+    void beginWritingOutputsForDstNodesInTable(common::table_id_t tableID) override;
+
+public:
+    BFSGraph bfsGraph;
+};
+
+class RJOutputWriter {
+public:
+    explicit RJOutputWriter(main::ClientContext* context, RJOutputs* rjOutputs);
+    virtual ~RJOutputWriter() = default;
+
+    void beginWritingForDstNodesInTable(common::table_id_t tableID) {
+        rjOutputs->beginWritingOutputsForDstNodesInTable(tableID);
+    }
+
+    virtual bool skipWriting(common::nodeID_t dstNodeID) const = 0;
+
+    virtual void write(processor::FactorizedTable& fTable, common::nodeID_t dstNodeID) const = 0;
+
+    virtual std::unique_ptr<RJOutputWriter> copy() = 0;
+
+protected:
+    main::ClientContext* context;
+    RJOutputs* rjOutputs;
+    std::unique_ptr<common::ValueVector> srcNodeIDVector;
+    std::unique_ptr<common::ValueVector> dstNodeIDVector;
+    std::vector<common::ValueVector*> vectors;
+};
+
+class DestinationsOutputWriter : public RJOutputWriter {
+public:
+    DestinationsOutputWriter(main::ClientContext* context, RJOutputs* rjOutputs)
+        : RJOutputWriter(context, rjOutputs){};
+
+    void write(processor::FactorizedTable& fTable, common::nodeID_t dstNodeID) const override;
+
+    bool skipWriting(common::nodeID_t dstNodeID) const override;
+
+    std::unique_ptr<RJOutputWriter> copy() override {
+        return std::make_unique<DestinationsOutputWriter>(context, rjOutputs);
+    }
+
+protected:
+    virtual void writeMoreAndAppend(processor::FactorizedTable& fTable, common::nodeID_t dstNodeID,
+        uint16_t length) const;
+};
+
+class PathsOutputWriter : public RJOutputWriter {
+public:
+    PathsOutputWriter(main::ClientContext* context, RJOutputs* rjOutputs, uint16_t lowerBound,
+        uint16_t upperBound);
+
+    void write(processor::FactorizedTable& fTable, common::nodeID_t dstNodeID) const override;
+
+protected:
+    uint16_t lowerBound;
+    uint16_t upperBound;
+    std::unique_ptr<common::ValueVector> lengthVector;
+    std::unique_ptr<common::ValueVector> pathNodeIDsVector;
+    std::unique_ptr<common::ValueVector> pathEdgeIDsVector;
+};
+
+class SPPathsOutputWriter : public PathsOutputWriter {
+public:
+    SPPathsOutputWriter(main::ClientContext* context, RJOutputs* rjOutputs, uint16_t upperBound)
+        : PathsOutputWriter(context, rjOutputs, 1 /* lower bound */, upperBound) {}
+
+    bool skipWriting(common::nodeID_t dstNodeID) const override {
+        auto pathsOutputs = rjOutputs->ptrCast<PathsOutputs>();
+        // For all shortest path computations, we do not output any results from source to source.
```

Comment:
Let's update this comment. This comment sounds like we don't output any results from s to s only for "all shortest paths" but we should not output s to s for "single shortest paths" as well. So let's be clear and say "For single or all shortest paths computations ..."

Reply:
Done